File: hamming_distance.py
########################
# author: Johnny Chang #
########################

## Description: Calculates Hamming Distance

def hamming_distance(x, y):
    diff_len = len(x) - len(y)     # take care of differences in string length
    hamming_dist = abs(diff_len)   # initialize Hamming distance holder with absolute value of differences in string length
    if diff_len < 0:               # if the length of x is smaller than y:
        for i in range(0, len(x)): # limit the comparison to the smaller length, in this case x
            if x[i] != y[i]:       # if the character at that index of x does not match with the character at that index of y, increase hamming distance by 1
                hamming_dist += 1
    else:                          # don't use another comparison, because the equal case is captured by len(y) as well
        for i in range(0, len(y)): # same as above
            if x[i] != y[i]:
                hamming_dist += 1

    return hamming_dist

print(hamming_distance("cat", "hat")) # expected: 1
print(hamming_distance("holy shit", "holy fuck")) # expected: 4
print(hamming_distance("gutenberg", "heisenberg")) # note that Levenshtein distance here would be a different story... tune in for next time

File: hamming_distance.c
#include <string.h>
#include <stdio.h>
#include <assert.h>

/*
char ** ARR_TEST_1 = {  "holy SHIT",
                        "",
                        "ARUBA",
                        "BITCHES",
                        "AMAZING",
                        "LOREM",
                        "KODALINE",
                        "KANYE WEST",
                        "holy fuck"}
*/

int hamming_distance(char * a, char * b) {
    int len_a = strlen(a);
    int len_b = strlen(b);

    /* take care of some edge cases */
    if (len_a == 0) {
        return len_b;
    } else if (len_b == 0) {
        return len_a;
    }

    int hamming_dist = len_a - len_b;
    if (hamming_dist < 0) { // this means that a is the smaller string
        while (*a) {        // while the character is not the NULL character
            if (*a != *b) { // increment Hamming distance if character differs
                hamming_dist++;
            }
            a++;            // increment the char we're looking at
            b++;
        }
    } else {                // this is also the len_a == len_b case
        while (*b) {
            if (*a != *b) {
                hamming_dist++;
            }
            a++;
            b++;
        }
    }

    return hamming_dist;
}

int main() {
    int ham_cat = hamming_distance("cat", "hat");
    printf("%d\n", ham_cat);
    //printf("'cat' and 'hat' Hamming distance is %d\n", ham_cat);
    //assert(hamming_distance("cat", "hat") == 1);

    int ham_holy = hamming_distance("holy shit", "holy fuck");
    printf("%d\n", ham_holy);

    int ham_berg = hamming_distance("heisenberg", "gutenberg");
    printf("%d\n", ham_berg);

    return 0;
}
